src/app/services/myinfo/myinfo.service.ts

```diff
@@ -0,0 +1,301 @@
+        const prefilledField = cloneDeep(field) as IPossiblyPrefilledField
+        prefilledField.fieldValue = myInfoValue
+
+        // Disable field
+        prefilledField.disabled = isReadOnly
+        return prefilledField
+      }),
+    )
+  }
+
+  /**
+   * Saves hashed prefilled values of MyInfo fields.
+   * @param prefilledFormFields Fields with fieldValue prefilled and disabled set to true if read-only
+   */
+  saveMyInfoHashes(
+    uinFin: string,
+    formId: string,
+    prefilledFormFields: IPossiblyPrefilledField[],
+  ): ResultAsync<IMyInfoHashSchema | null, HashingError | DatabaseError> {
+    const readOnlyHashPromises = createHashPromises(prefilledFormFields)
+    return ResultAsync.fromPromise(
+      Bluebird.props<IHashes>(readOnlyHashPromises),
+      (error) => {
+        const message = 'Failed to hash MyInfo values'
+        logger.error({
+          message,
+          meta: {
+            action: 'saveMyInfoHashes',
+            myInfoAttributes: Object.keys(readOnlyHashPromises),
+          },
+          error,
+        })
+        return new HashingError(message)
+      },
+    ).andThen((readOnlyHashes: IHashes) => {
+      return ResultAsync.fromPromise(
+        MyInfoHash.updateHashes(
+          uinFin,
+          formId,
+          readOnlyHashes,
+          this.#spCookieMaxAge,
+        ),
+        (error) => {
+          const message = 'Failed to save MyInfo hashes to database'
+          logger.error({
+            message,
+            meta: {
+              action: 'saveMyInfoHashes',
+              myInfoAttributes: Object.keys(readOnlyHashPromises),
+            },
+            error,
+          })
+          return new DatabaseError(message)
+        },
+      )
+    })
+  }
+
+  fetchMyInfoHashes(
+    uinFin: string,
+    formId: string,
+  ): ResultAsync<IHashes, DatabaseError | MissingHashError> {
+    return ResultAsync.fromPromise(
+      MyInfoHash.findHashes(uinFin, formId),
+      (error) => {
+        const message = 'Error while fetching MyInfo hashes from database'
+        logger.error({
+          message,
+          meta: {
+            action: 'fetchMyInfoHashes',
+          },
+          error,
+        })
+        return new DatabaseError(message)
+      },
+    ).andThen((hashes) => {
+      if (hashes) {
+        return okAsync(hashes)
+      } else {
+        logger.info({
+          message: 'MyInfo hashes expired',
+          meta: {
+            action: 'fetchMyInfoHashes',
+            formId,
+          },
+        })
+      }
+      return errAsync(new MissingHashError('MyInfo hashes not found.'))
+    })
+  }
+
+  checkMyInfoHashes(
+    responses: ProcessedFieldResponse[],
+    hashes: IHashes,
+  ): ResultAsync<IHashes, HashingError | HashDidNotMatchError> {
+    // Filter twice to get the types to cooperate
+    const responsesWithHashes: VisibleMyInfoResponse[] = responses
+      .filter(hasMyInfoAnswer)
+      .filter((response) => !!hashes[response.myInfo.attr])
```

Comment:
i believe you won't have to filter twice if you convert `hasMyInfoAnswer` into a type guard

Reply:
wait, the function is already a type guard; this filter is to obtain the responses that have had their hashes saved.

Reply:
it already is a type guard! apparently using a filter function like `(element) => typeGuard(element) && otherFunction(element)` doesn't actually return the correct type...